pkg/connector/kairosdb.go

```diff
@@ -0,0 +1,494 @@
+// +build kairosdb
+
+package connector
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/facette/facette/pkg/catalog"
+	"github.com/facette/facette/pkg/config"
+	"github.com/facette/facette/pkg/logger"
+	"github.com/facette/facette/pkg/plot"
+	"github.com/facette/facette/pkg/utils"
+)
+
+const (
+	kairosdbURLVersion     string  = "/api/v1/version"
+	kairosdbURLMetricNames string  = "/api/v1/metricnames"
+	kairosdbURLMetricsTags string  = "/api/v1/datapoints/query/tags"
+	kairosdbURLQueryMetric string  = "/api/v1/datapoints/query"
+	kairosdbDefaultTimeout float64 = 10
+)
+
+type kairosdbSeriesEntry struct {
+	metric     string
+	tag        string
+	source     string
+	aggregator MetricsQueryAggregation
+}
+
+type MetricsQueryAggregation struct {
+	Name     string            `json:"name"`
+	Sampling map[string]string `json:"sampling"`
+}
+
+type MetricsQueryEntry struct {
+	Name        string                    `json:"name"`
+	Tags        map[string][]string       `json:"tags"`
+	Aggregators []MetricsQueryAggregation `json:"aggregators"`
+}
+
+type MetricsQueryResponse struct {
+	Sample_size int64                `json:"sample_size"`
+	Results     []MetricsQueryResult `json:"results"`
+}
+
+type MetricsQueryResult struct {
+	Name     string              `json:"name"`
+	Group_by []map[string]string `json:"group_by"`
+	Tags     map[string][]string `json:"tags"`
+	Values   [][2]float64        `json:"values"`
+}
+
+// KairosdbConnector represents the main structure of the Kairosdb connector.
+type KairosdbConnector struct {
+	name        string
+	URL         string
+	insecureTLS bool
+	timeout     float64
+	srcTags     []string       // TODO: make configurable
+	sourceRe    *regexp.Regexp // REVIEW: see below
+	metricRe    *regexp.Regexp // REVIEW: see below
+	series      map[string]map[string]kairosdbSeriesEntry
+}
+
+func init() {
+	Connectors["kairosdb"] = func(name string, settings map[string]interface{}) (Connector, error) {
+		// var sp, mp string // REVIEW: see below
+		var err error
+
+		connector := &KairosdbConnector{
+			name:        name,
+			insecureTLS: true,
+			series:      make(map[string]map[string]kairosdbSeriesEntry),
+			srcTags:     []string{"host", "name"}, // TODO: make configurable
```

Comment:
Am 18.01.2015 um 15:53 schrieb Marc Falzon:
> In pkg/connector/kairosdb.go
> <https://github.com/facette/facette/pull/153#discussion-diff-23133775>:
> 
>> +			srcTags:     []string{"host", "name"}, // TODO: make configurable
> 
> I'm afraid I'm not competent enough with KairosDB at the moment, I don't
> know how to use other tags than |host|. Could you suggest a use case
> where other tags could be used in the context of Facette?

This is because metrics are not standardized and any application
that is feeding data to a backend (KairosDB in this case) have its
own meaning about what is the right tag for the thing called 'source'
by facette.

We are feeding data via collectd's Graphite channel to KairosDB.
The sources are tagged as 'host' here. But we are monitoring
VCenters with an app called statsfeeder. The sources are tagged
as 'name' here.

Hoping that http://metrics20.org/ helps to overcome this in the
next years.

I'll try to make it configurable until next commit.